src/main/java/com/hrznstudio/titanium/recipe/generator/titanium/ResourceRegistryProvider.java

```diff
@@ -0,0 +1,92 @@
+        this.generator = generator;
+        this.itemTagsProvider = new ItemTagsProvider(generator) {
+            @Override
+            protected void registerTags() {
+                ResourceRegistry.getMaterials().forEach(material -> {
+                    material.getGenerated().values().stream().filter(entry -> entry instanceof IResourceHolder).forEach(entry -> {
+                        if (entry instanceof BlockBase) {
+                            this.copy(new BlockTags.Wrapper(new ResourceLocation("forge", ((IResourceHolder) entry).getType().getTag() + "/" + ((IResourceHolder) entry).getMaterial().getMaterialType())),
+                                    new ItemTags.Wrapper(new ResourceLocation("forge", ((IResourceHolder) entry).getType().getTag() + "/" + ((IResourceHolder) entry).getMaterial().getMaterialType())));
+                        } else if (entry instanceof ItemBase) {
+                            this.getBuilder(new ItemTags.Wrapper(new ResourceLocation("forge", ((IResourceHolder) entry).getType().getTag() + "/" + ((IResourceHolder) entry).getMaterial().getMaterialType()))).add((Item) entry);
+                        }
+                    });
+                });
+            }
+        };
+        this.blockTagsProvider = new BlockTagsProvider(generator) {
+            @Override
+            protected void registerTags() {
+                ResourceRegistry.getMaterials().forEach(material -> {
+                    material.getGenerated().values().stream().filter(entry -> entry instanceof IResourceHolder).forEach(entry -> {
+                        if (entry instanceof BlockBase) {
+                            this.getBuilder(new BlockTags.Wrapper(new ResourceLocation("forge", ((IResourceHolder) entry).getType().getTag() + "/" + ((IResourceHolder) entry).getMaterial().getMaterialType()))).add((Block) entry);
+                        }
+                    });
+                });
+            }
+        };
+    }
+
+    @Override
+    public void act(DirectoryCache cache) throws IOException {
+        this.itemTagsProvider.act(cache);
+        this.blockTagsProvider.act(cache);
+        Path path = this.generator.getOutputFolder();
+        ResourceRegistry.getMaterials().forEach(material -> {
+            material.getGenerated().values().stream().filter(entry -> entry instanceof IJSONGenerator && entry instanceof IJsonFile).forEach(entry -> {
+                Path output = path.resolve((((IJsonFile) entry).getRecipeSubfolder() != null ? ((IJsonFile) entry).getRecipeSubfolder() : ""));
+                try {
+                    Files.createDirectories(output);
+                    try (BufferedWriter bufferedwriter = Files.newBufferedWriter(output.resolve(((IJsonFile) entry).getRecipeKey() + ".json"))) {
+                        bufferedwriter.write(GSON.toJson(((IJSONGenerator) entry).generate()));
+                    }
+                } catch (IOException e) {
+                    e.printStackTrace();
```

Comment:
Log don't print.

Reply:
This exception doesn't run when mc is open so a proper stacktrace is needed just in case

Reply:
What I said sounds dumb ignore me